src/Workspaces/CSharp/Portable/CodeGeneration/CSharpCodeGenerationHelpers.cs

```diff
@@ -21,6 +20,67 @@ namespace Microsoft.CodeAnalysis.CSharp.CodeGeneration
 {
     internal static class CSharpCodeGenerationHelpers
     {
+        public static void AdjustNullableAnnotationByAttributes(ITypeSymbol type, ref ImmutableArray<AttributeData> attributes, ref NullableAnnotation nullableAnnotation, bool isParameter)
+        {
+            if (type.IsReferenceType)
+            {
+                if (!(type is ITypeParameterSymbol typeParameter) ||
+                    typeParameter.ReferenceTypeConstraintNullableAnnotation == NullableAnnotation.NotAnnotated)
+                {
+                    // For [AllowNull]/[MaybeNull], add the nullable annotation.
+                    // ([AllowNull] string?) -> (string?)
+                    // ([AllowNull] string)  -> (string?)
```

Comment:
can we add comments exaplining the TYpeparameter + ReferenceTypeConstraint case?